### ISSUE TITLE
braces needed around crossref field

### ISSUE THREAD

**Issue by ralessi** [joostkremers/ebib#123]
At least with the latest texlive installation, `biber` complains when it does not find braces around the `crossref` field, and in the resulting bibliography, the cross-referenced data is discarded.

Is there a way to have braces automatically inserted by ebib around this field?

Many thanks for this *excellent* package! -- Robert

**Comment by joostkremers** [comment]
This is a `.bib` file that wasn't created with Ebib, I assume? Because normally Ebib will put braces around all field values unless you explicitly tell it not to.

You should see an asterisk in front of the `crossref` value, which tells you this field does not have braces. You can change this with the command `ebib-toggle-raw`, which is bound to `r` in the entry buffer.

Obviously that's rather cumbersome if you have a lot of entries to go through. Currently, there is no automatic way to add braces to all occurrences of a specific field automatically, however.

It also wouldn't be correct to have Ebib add braces around the `crossref` field automatically, because it's possible to use macros (defined with `@string`) in the `crossref` field and braces would make that impossible.

You can do a regexp replace in the actual `.bib` files, or, if you're adventurous, execute the following Elisp snippet in the `*scratch*` buffer after opening Ebib, loading all your `.bib` files and lowering Ebib with `z`:

```
(mapc (lambda (db)
        (mapc (lambda (key)
                (ebib-set-field-value "crossref"
                                      (ebib-get-field-value "crossref" key db 'noerror)
                                      key db 'overwrite))
              (ebib-db-list-keys db))
        ebib--databases))
```

I say "adventurous" because I haven't actually tested it. But it should go through your databases and add braces around any `crossref` value in them. After executing it, return to Ebib and save all databases (there's a command for this in the menu). Do back up your `.bib` files if you try this. :smiley:

**Comment by ralessi** [comment]
> Hitting the `r` key produced the same result.

Maybe I should add that hitting the `r` key on an empty `crossref` field doesn't make Ebib ask me for a value.  Just nothing happens.

**Comment by joostkremers** [comment]
> I also tried to start with a new file. I created only two entries: one being of `Book` type and the other of `BookInBook`. And when I tried to crossref the latter to the former, I got the asterisk in front of the crossref field. Is that the expected behavior? 

No, it's not. That's one thing that surprises me.

> Hitting the `r` key produced the same result. 

Looking at the code, I just realised that hitting `r` on the `crossref` field doesn't do anything. I had forgotten that... 

> So I suspect that something might be wrong with my installation of Ebib, am I correct in assuming this?

It's possible, but I have no idea what could be wrong... Did you make any customisations for Ebib? I assume you have the latest version installed?

> Hitting the `r` key produced the same result.
> Maybe I should add that hitting the `r` key on an empty `crossref` field doesn't make Ebib ask me for a value. Just nothing happens.

That's actually the way it's currently implemented, so that's not surprising. But given that macros can appear in the `crossref` field, it's obviously wrong and should be changed.

So to recap: the only thing that doesn't seem to work the way it should is that when you edit the `crossref` field, you get an unbraced value. That shouldn't be happening.

**Comment by ralessi** [comment]
> > So I suspect that something might be wrong with my installation of Ebib, am I correct in assuming this?
> 
> It's possible, but I have no idea what could be wrong... Did you make any customisations for Ebib? I assume you have the latest version installed?

I have 2.14.1.  Otherwise, I did not customize it much.  Here is what I did:

````
 '(ebib-autogenerate-keys t)
 '(ebib-bib-search-dirs (quote ("~/texmf/bibtex/bib" "~")))
 '(ebib-bibtex-dialect (quote biblatex))
 '(ebib-extra-fields
   (quote
    ((biblatex "crossref" "annote" "abstract" "keywords" "file" "timestamp" "url" "doi")
     (biblatex "crossref" "annotation" "abstract" "keywords" "file" "timestamp"))))
 '(ebib-file-associations (quote (("pdf" . "zathura") ("ps" . "gv"))))
 '(ebib-index-column-separator " | ")
 '(ebib-index-columns
   (quote
    (("Entry Key" 15 t)
     ("Author/Editor" 15 t)
     ("Year" 6 t)
     ("Title" 40 t))))
 '(ebib-preload-bib-files (quote ("<pathto>/Library/index.bib")))

````

I also tried to remove those lines, I even tried to move .emacs to .emacs-out but no avail.

> 
> > Hitting the `r` key produced the same result.
> > Maybe I should add that hitting the `r` key on an empty `crossref` field doesn't make Ebib ask me for a value. Just nothing happens.
> 
> That's actually the way it's currently implemented, so that's not surprising. But given that macros can appear in the `crossref` field, it's obviously wrong and should be changed.
> 
> So to recap: the only thing that doesn't seem to work the way it should is that when you edit the `crossref` field, you get an unbraced value. That shouldn't be happening.

That is correct, in addition to hitting `r` on an empty `crossref` producing nothing.

Many thanks for your help!

**Comment by joostkremers** [comment]
Well, here's me being silly. The third most recent commit has the subject line "Bug fix: add braces when editing the crossref field." It's only a few weeks old (December 21), but I had already forgotten about it.

So this is actually a bug that's already fixed. I've just now added a tag (2.14.2) to the relevant commit, so that it'll also show up on Melpa Stable, in case you're using that. Once you update Ebib, your issue should be solved.

>  in addition to hitting `r` on an empty `crossref` producing nothing. 

I'll fix that some time in the next few weeks.

**Comment by ralessi** [comment]
Melpa just gave me 2.14.3, which does write braces around `crossref`, many thanks!  However, the `save database` function does not seem to work anymore: hitting `s` gives me what follows in the mini-buffer: `Symbol's function definition is void: when-let*`

This latest version also compiled with some errors.  Compile-Log follows.  Should I try to remove Ebib and reinstall it?

````
Leaving directory ‘/home/robert/.emacs.d/elpa/ebib-2.14.3’

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/ebib-db.el at Sun Jan  6 10:23:19 2019
Entering directory ‘/home/robert/.emacs.d/elpa/ebib-2.14.3/’

In end of data:
ebib-db.el:504:1:Warning: the function ‘hash-table-keys’ is not known to be
    defined.

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/ebib-filters.el at Sun Jan  6 10:23:19 2019

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/ebib-keywords.el at Sun Jan  6 10:23:19 2019

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/ebib-notes.el at Sun Jan  6 10:23:19 2019

In end of data:
ebib-notes.el:261:1:Warning: the function ‘ebib-lower’ is not known to be
    defined.

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/ebib-reading-list.el at Sun Jan  6 10:23:19 2019

In end of data:
ebib-reading-list.el:216:1:Warning: the function ‘org-todo’ is not known to be
    defined.

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/ebib-utils.el at Sun Jan  6 10:23:19 2019

In end of data:
ebib-utils.el:1808:1:Warning: the following functions are not known to be defined:
    ebib--set-modified, ebib--format-entry

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/ebib.el at Sun Jan  6 10:23:19 2019

In ebib-index-mode:
ebib.el:769:45:Warning: assignment to free variable ‘hl-line-face’
ebib.el:1477:1:Error: Wrong number of arguments: (3 . 5), 6

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.3/org-ebib.el at Sun Jan  6 10:23:20 2019
````

**Comment by joostkremers** [comment]
Oh... Some of those warnings mean that I've used functions that are too new, specifically `when-let*` and `hash-table-keys`. I'll push a fix to remove those, will let you know when it's done.

The other warnings are (hopefully) caused by the earlier ones, because some parts of the files don't compile correctly.

Unfortunately, where I am right now I only have Emacs 26 to test my code against, so it's a bit difficult to make sure it works on 25 as well... :cry:

**Comment by joostkremers** [comment]
Well, as it turns out, both `when-let*' and `hash-table-keys' exist in Emacs 25. The problem was that the library that defines them wasn't `require`d in the Ebib source files that use them. That should be fixed now. I hope that also fixes your other compiler warnings...

**Comment by ralessi** [comment]
> Well, as it turns out, both `when-let*' and `hash-table-keys' exist in Emacs 25. The problem was that the library that defines them wasn't `require`d in the Ebib source files that use them. That should be fixed now. I hope that also fixes your other compiler warnings...

Thank you again!  As a matter of fact, I have emacs 25.3.1.  Unfortunately, installing Ebib 2.14.5 gave me the log below, and I still have the warning related to `when-let*` when I hit the `s` key.

Do you think that installing emacs 26 would solve this issue?

````

Leaving directory ‘/home/robert/.emacs.d/elpa/parsebib-2.3.2’

Compiling file /home/robert/.emacs.d/elpa/parsebib-2.3.2/parsebib.el at Mon Jan  7 08:54:28 2019
Entering directory ‘/home/robert/.emacs.d/elpa/parsebib-2.3.2/’

Compiling no file at Mon Jan  7 08:54:30 2019
Leaving directory ‘/home/robert/.emacs.d/elpa/ebib-2.14.5’

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib-db.el at Mon Jan  7 08:54:30 2019
Entering directory ‘/home/robert/.emacs.d/elpa/ebib-2.14.5/’

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib-filters.el at Mon Jan  7 08:54:30 2019

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib-keywords.el at Mon Jan  7 08:54:31 2019

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib-notes.el at Mon Jan  7 08:54:31 2019

In end of data:
ebib-notes.el:261:1:Warning: the function ‘ebib-lower’ is not known to be
    defined.

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib-reading-list.el at Mon Jan  7 08:54:31 2019

In end of data:
ebib-reading-list.el:216:1:Warning: the function ‘org-todo’ is not known to be
    defined.

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib-utils.el at Mon Jan  7 08:54:31 2019
ebib-utils.el:1433:1:Error: Wrong number of arguments: (3 . 5), 6

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib.el at Mon Jan  7 08:54:31 2019

In ebib-index-mode:
ebib.el:768:45:Warning: assignment to free variable ‘hl-line-face’
ebib.el:2354:1:Warning: Unused lexical variable ‘ivy-sort-max-size’

In ebib-ivy-insert-citation:
ebib.el:2365:11:Warning: assignment to free variable ‘ivy-completion-beg’
ebib.el:2366:11:Warning: assignment to free variable ‘ivy-completion-end’
ebib.el:2372:48:Warning: reference to free variable ‘ivy-completion-beg’
ebib.el:2372:67:Warning: reference to free variable ‘ivy-completion-end’

In ebib-entry-mode:
ebib.el:2703:45:Warning: assignment to free variable ‘hl-line-face’

In ebib-strings-mode:
ebib.el:3192:45:Warning: assignment to free variable ‘hl-line-face’

In end of data:
ebib.el:3725:1:Warning: the following functions are not known to be defined:
    hl-line-highlight, ivy-read, with-ivy-window,
    hl-line-unhighlight

Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/org-ebib.el at Mon Jan  7 08:54:31 2019
````

**Comment by joostkremers** [comment]
> Thank you again! As a matter of fact, I have emacs 25.3.1. Unfortunately, installing Ebib 2.14.5 gave me the log below, 

Most of those messages are actually harmless.

> and I still have the warning related to `when-let*` when I hit the `s` key.

Well, I just found out that Emacs 25 has `when-let`, but not `when-let*`, so that explains it.

The other problem is this:

> Compiling file /home/robert/.emacs.d/elpa/ebib-2.14.5/ebib-utils.el at Mon Jan 7 08:54:31 2019 ebib-utils.el:1433:1:Error: Wrong number of arguments: (3 . 5), 6

The error means that some function is called with six arguments, although it accepts at most five. This may happen when an existing function is given an extra argument in a newer Emacs version. Unfortunately, the error doesn't mention the offending function and the file+line number don't help, because there is no function with six arguments on that line or even close to it. So I really need to find a way to install a working copy of Emacs 25 myself...

> Do you think that installing emacs 26 would solve this issue?

That should certainly solve it, because I run Emacs 26 myself and I don't have the problems you experience. But I'd still like to maintain compatibility with Emacs 25, so I'll need to figure out what's going wrong anyway.

**Comment by joostkremers** [comment]
Ok, the Emacs 25 incompatibilities should now be fixed. I managed to install Emacs 25 in a VirtualBox VM and it compiles without serious issues. There are still a few compilation warnings, but they shouldn't pose any problems.

Do let me know if you run into further trouble, though... :anguished:

**Comment by ralessi** [comment]
> Well, as it turns out, both `when-let*' and `hash-table-keys' exist in Emacs 25. The problem was that the library that defines them wasn't `require`d in the Ebib source files that use them. That should be fixed now. I hope that also fixes your other compiler warnings...

As far as I can tell, this issue is completely fixed: I could install the latest Ebib without getting any warning, just I used to install the previous releases.  The crossref field is now surrounded with braces, and the `s` key does its job as normal.

In one word: bravo! -- From a very grateful user

**Comment by joostkremers** [comment]
Great! That's good to hear. :smiley:
